packages/attestation-service/src/sms/index.ts

```diff
@@ -17,31 +26,242 @@ export async function initializeSmsProviders() {
+    }
+
+    // If this provider supports delivery status, track the delivery.
+    if (delivery.providers[0].supportsDeliveryStatus()) {
+      ongoingDeliveries[deliveryId!] = delivery
+      delivery.deliveryId = deliveryId
+      delivery.status = DeliveryStatus.Created
+
+      // Set timeout to clean up in event of no delivery receipt
+      setTimeout(() => {
+        if (delivery.deliveryId === deliveryId) {
+          delete ongoingDeliveries[deliveryId]
+          if (delivery.finallyBelievedDeliveredCallback) {
+            delivery.finallyBelievedDeliveredCallback()
+          }
+          delivery.deliveryId = null
+        }
+      }, timeoutWaitingForDeliveryReceipt)
+    }
+    return delivery.providers[0].type
+  } catch (error) {
+    rootLogger.info(
+      {
+        provider: delivery.providers[0].type,
+        phoneNumber: obfuscateNumber(delivery.phoneNumber),
+        attempt: delivery.attemptsForThisProvider,
+        error,
+      },
+      'SMS creation failed'
+    )
+
+    // Error sending. Set timeout to backoff and retry
+    await sleep(Math.pow(2, delivery.attemptsForThisProvider) * 1000)
+    return attemptToSendSms(delivery)
+  }
+}
+
+// Timeout waiting for a delivery receipt (may never come, even on success) before delivery state is cleaned up
+const timeoutWaitingForDeliveryReceipt =
+  parseInt(fetchEnvOrDefault('TIMEOUT_CLEANUP_NO_RECEIPT_MIN', '10'), 10) * 60 * 1000
+
+// Act on a delivery report for an SMS, scheduling a resend if last one failed.
+export function receivedDeliveryReport(
+  deliveryId: string,
+  deliveryStatus: DeliveryStatus,
+  errorCode: string | null
+) {
+  const delivery = ongoingDeliveries[deliveryId]
+  if (delivery) {
+    rootLogger.info(
+      {
+        provider: delivery.providers[0].type,
+        phoneNumber: obfuscateNumber(delivery.phoneNumber),
+        deliveryId,
+        deliveryStatus: DeliveryStatus[deliveryStatus],
+        errorCode,
+      },
+      'Received delivery status'
+    )
+
+    if (delivery.status !== deliveryStatus) {
+      delivery.status = deliveryStatus
+
+      Counters.attestationProviderDeliveryStatus
+        .labels(delivery.providers[0].type, delivery.countryCode, DeliveryStatus[deliveryStatus])
+        .inc()
+
+      if (errorCode != null) {
+        Counters.attestationProviderDeliveryStatus
+          .labels(delivery.providers[0].type, delivery.countryCode, errorCode)
+          .inc()
+      }
+    }
+
+    if (deliveryStatus === DeliveryStatus.Delivered) {
+      delete ongoingDeliveries[deliveryId]
+      ongoingDeliveries.deliveryId = null
+      if (delivery.finallyBelievedDeliveredCallback) {
+        delivery.finallyBelievedDeliveredCallback()
+      }
+    } else if (deliveryStatus === DeliveryStatus.Failed) {
+      delete ongoingDeliveries[deliveryId]
+      ongoingDeliveries.deliveryId = null
+      setTimeout(() => {
+        void attemptToSendSms(delivery)
+      }, Math.pow(2, delivery.attemptsForThisProvider) * 3000)
```

Comment:
something to note in a comment maybe, that since this is exponential, it will surpass  the wait time for the `finallyBelievedDeliveredCallback` timeout at around the 8th attempt (since the attempt amount is configurable)

Reply:
They're two separate timeouts... the `finallyBelievedDeliveredCallback` only kicks in after a successful send, when we give up waiting for a delivery receipt reporting success and assume it's delivered ok anyway.  This timeout only kicks in after an attempt sending the SMS fails, before we retry. At that point no SMS is in flight.

Reply:
got it 👍

Reply:
thanks for explaining -- overall these function may benefit from breaking down each step into a separate function for future eng readability.